agent/rpc/subscribe/subscribe.go

```diff
@@ -0,0 +1,212 @@
+	authz, err := h.Backend.ResolveToken(req.Token)
+	if err != nil {
+		return err
+	}
+
+	sub, err := h.Backend.Subscribe(toStreamSubscribeRequest(req))
+	if err != nil {
+		return err
+	}
+	defer sub.Unsubscribe()
+
+	ctx := serverStream.Context()
+
+	elog := &eventLogger{logger: logger}
+	for {
+		events, err := sub.Next(ctx)
+		switch {
+		case errors.Is(err, stream.ErrSubscriptionClosed):
+			logger.Trace("subscription reset by server")
+			return status.Error(codes.Aborted, err.Error())
+		case err != nil:
+			return err
+		}
+
+		events = filterStreamEvents(authz, events)
+		if len(events) == 0 {
+			continue
+		}
+
+		elog.Trace(events)
+		e := newEventFromStreamEvents(req, events)
+		if err := serverStream.Send(e); err != nil {
+			return err
+		}
+	}
+}
+
+// TODO: can be replaced by mog conversion
+func toStreamSubscribeRequest(req *pbsubscribe.SubscribeRequest) *stream.SubscribeRequest {
+	return &stream.SubscribeRequest{
+		Topic: req.Topic,
+		Key:   req.Key,
+		Token: req.Token,
+		Index: req.Index,
+	}
+}
+
+func forwardToDC(
+	req *pbsubscribe.SubscribeRequest,
+	serverStream pbsubscribe.StateChangeSubscription_SubscribeServer,
+	logger Logger,
+) func(conn *grpc.ClientConn) error {
+	return func(conn *grpc.ClientConn) error {
+		logger.Trace("forwarding to another DC")
+		defer logger.Trace("forwarded stream closed")
+
+		client := pbsubscribe.NewStateChangeSubscriptionClient(conn)
+		streamHandle, err := client.Subscribe(serverStream.Context(), req)
+		if err != nil {
+			return err
+		}
+
+		for {
+			event, err := streamHandle.Recv()
+			if err != nil {
+				return err
+			}
+			if err := serverStream.Send(event); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// filterStreamEvents to only those allowed by the acl token.
+func filterStreamEvents(authz acl.Authorizer, events []stream.Event) []stream.Event {
+	// authz will be nil when ACLs are disabled
+	if authz == nil || len(events) == 0 {
+		return events
+	}
+
+	// Fast path for the common case of only 1 event since we can avoid slice
+	// allocation in the hot path of every single update event delivered in vast
+	// majority of cases with this. Note that this is called _per event/item_ when
+	// sending snapshots which is a lot worse than being called once on regular
+	// result.
+	if len(events) == 1 {
+		if enforceACL(authz, events[0]) == acl.Allow {
+			return events
+		}
+		return nil
+	}
+
+	var filtered []stream.Event
+	for idx := range events {
+		event := events[idx]
+		if enforceACL(authz, event) == acl.Allow {
+			filtered = append(filtered, event)
+		}
+	}
+	return filtered
+}
+
+func newEventFromStreamEvents(req *pbsubscribe.SubscribeRequest, events []stream.Event) *pbsubscribe.Event {
+	e := &pbsubscribe.Event{
+		Topic: req.Topic,
+		Key:   req.Key,
+		Index: events[0].Index,
+	}
+
+	if len(events) == 1 {
+		event := events[0]
+		// TODO: refactor so these are only checked once, instead of 3 times.
+		switch {
```

Comment:
Why not handle these cases in switch within `setPayload`?

Reply:
I like that idea. I don't think it addresses the TODO, but it should remove one extra switch

Reply:
I tried it out. The reason reason this doesn't work is that the types for the framing events are not exported. So we need separate switches either way. One is a type switch and the other uses no clause and the cases themselves evaluate to true/false.

Reply:
Sounds good to me.